### PR TITLE
[v3] Prepare for v3

### PR DESCRIPTION
**Pull request by ErikSchierboom**
_This issue is part of the migration to v3. You can read full details about the various changes [here](https://github.com/exercism/v3-launch/)._

This PR prepares the track for Exercism v3, which will be different in a number of ways from Exercism v2.

In this PR, the following changes are made:

## v2 file migration

1. Move the existing exercises in the `exercises` directory to the `exercises/practice` directory

## v3 file migration

1. Copy the v3 `languages/<slug>/concepts` directory to the `concepts` directory
1. Copy the v3 `languages/<slug>/exercises/concept` directory to the `exercises/concept` directory
1. Copy the v3 `languages/<slug>/reference` directory to the `reference` directory
1. Copy the v3 `languages/<slug>/docs` directory to the `docs` directory

### Notes

Any commits modifying the aforementioned directories (or any of their files) are included in the PR, except for:

- The `_sidebar.md` files are not migrated
- The `README.md` file at the v3 track's root is not migrated.
- The `maintainers.md` file is not migrated.

## config.json migration

1. Add a version property:

```json
"version": 3
```

2. Add online editor settings:

```json
"online_editor": {
  "indent_style": "space",
  "indent_size": 2
}
```

3. Rename the `"exercises"` property to `"practice"`

4. Create an `"exercises"` property and move the `"practice"` property to this property:

```json
"exercises": {
  "practice": [
    ...
  ]
}
```

5. Remove the `"core"`, `"auto_approve"`, and `"unlocked_by"` properties from the practice exercises

6. Add the `"name"` property to the practice exercises and pre-populate this with a titlelized version of the `"slug"` property

7. Add the `"prerequisites"` property to the practice exercises and set it to an empty array

8. Add an empty `"concept"` array property to the `"exercises"` property

```json
"exercises": {
  "concept": [],
  "practice": [
    ...
  ]
}
```

9. Move the `"foregone"` property to the `"exercises"` key:

```json
"exercises": {
  ...
  "foregone": [...]
}
```

10. Add a top-level `"concepts"` key, which is an array:

```json
"concepts": []
```

11. Add a top-level `"tags"` key, which is an array:

```json
"tags": []
```

12. Add a top-level `"key_features"` key, which is an array:

```json
"key_features": []
```

13. Add a top-level `"v3_ready"` key, which is a boolean value that is set to `true` if the track has a `config.json` file in the v3 repo, and otherwise set to `false`:

```json
"v3_ready": true
```

14. Add a top-level `"slug"` key, which is a string containing the track's slug:

```json
"slug": "csharp"
```

### Notes

- Settings defined in the v3 `config.json` file will take precedence over their v2 `config.json` file's equivalent.

## configlet

1. Update the `fetch-configlet` and `fetch-configlet.ps1` files to the [latest version of configlet (v3)](https://github.com/exercism/canonical-data-syncer), which can work with v3 tracks.

## Continuous integration

1. Add a GitHub Actions workflow to verify the track using the [configlet-CI GitHub Action](https://github.com/exercism/github-actions/tree/master/configlet-ci), unless there already is a file named `.github/workflows/configlet.md`.

1. Add a [dependabot configuration](https://dependabot.com/github-actions/) to automatically submit PRs for any new versions of external workflows used in this track's GitHub Action workflows.

## Follow-up steps

We've created issues in this repo for the follow-up steps to get this track ready for v3.

### CODE REVIEW

**Review by coriolinus — COMMENTED** [review]
- Moving all existing `exercises/E/*` to `exercises/practice/E/*` is going to break current track scripting/CI.
- I don't remember what the `reference` folder is for. Most documents within it appear stale. If there's no canonical use for it, I'd prefer to just not import it.
- It appears that the merge script has added you as committer to every commit imported. That's fine, but IMO it would be nicer to retain the original commit attribution.

Is there a link to comprehensive documentation about the new track structure somewhere? That will be very handy when I have to fix everything that breaks.

File: .github/workflows/configlet.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+name: Configlet CI
```
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply .github/workflows/configlet.yml:1]: Correct. Once the PR has been merged, tracks can decide for themselves what the best approach is.

File: config.json [hunk 2]
```diff
@@ -1,1178 +1,1325 @@
 {
   "language": "Rust",
+  "slug": "rust",
   "active": true,
+  "v3_ready": true,
```
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply config.json:5]: Rust _does_ have a test runner, right? That's for now the main difference between v3-ready and non-ready tracks.

[the remark this answers is not in the document]
Comment (coriolinus) [reply config.json:5]: It does, but I'd prefer to defer launching v3 until the concept exercises are looking good. Right now, they're not even minimal. That's a different discussion though.

**Comment by ErikSchierboom** [comment]
> Moving all existing exercises/E/* to exercises/practice/E/* is going to break current track scripting/CI.


> I don't remember what the reference folder is for. Most documents within it appear stale. If there's no canonical use for it, I'd prefer to just not import it.

It can serve different functions per tracks. Most tracks have used it for the "extract concepts from v2 exercises" step that tracks engaged in at the beginning of the v3 repo. Other tracks have added documentation to this directory. In essence: tracks are free to do whatever they want with this directory. My script just copies it. If you (the Rust maintainers) decide that it doesn't serve any purpose, the idea is that you can remove it after the PR has been merged. I'd rather not make track-specific exceptions.

> It appears that the merge script has added you as committer to every commit imported. That's fine, but IMO it would be nicer to retain the original commit attribution.

It would be nicer, but I've not yet found how to do that. I'm using this tool for the history import: https://htmlpreview.github.io/?https://github.com/newren/git-filter-repo/blob/docs/html/git-filter-repo.html

> Is there a link to comprehensive documentation about the new track structure somewhere? That will be very handy when I have to fix everything that breaks.

There is. From now on, all documentation lives in the [v3-docs repo](https://github.com/exercism/v3-docs). You'll be particularly interested in https://github.com/exercism/v3-docs/tree/main/anatomy/tracks, which contains documents for the specifications of config.json files and similar files.

**Comment by coriolinus** [comment]
> I'd rather not make track-specific exceptions.

Yeah, that's totally fair.

Thanks for the link to the documentation!

File: config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply config.json:5]: True.

File: reference/merge-strategy.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# merging strategy
```
Comment (efx) [inline reference/merge-strategy.md:1]: We'll be able to condense this file into a general CONTRIBUTING document when the time is right.

File: config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (efx) [reply config.json:5]: The core structure is ready but I agree with @coriolinus that we should set this to false now. We'll be coordinating sometime this week about how to implement complete concept exercises better in the coming weeks.

**Comment by efx** [comment]
@ErikSchierboom did you intend to target your fork of the exercism/rust? Or should this target exercism/rust?
I'm curious if we could create the main branch in exercism/rust and target that for all changes in advance of the cutover on Friday.

**Comment by ErikSchierboom** [comment]
@efx I'm sorry, but that won't be possible, as my script has to work with the latest version and to prevent any issues, I'm resetting branches each time I run the script. You could theoretically base your branches of this PR's branch, you'd just have to add a different remote.

**Pull request closed without merging**